### PR TITLE
Throw error if scalar template type doesn't match array type

### PR DESCRIPTION
**Pull request by 9prady9**
Related to #1949

### CODE REVIEW

File: src/api/cpp/array.cpp [hunk 1]
```diff
@@ -980,6 +980,10 @@ af::dtype implicit_dtype(af::dtype scalar_type, af::dtype array_type)
     }                                                               \
     template<> AFAPI T array::scalar() const                        \
     {                                                               \
+        af_dtype type = (af_dtype)af::dtype_traits<T>::af_type;     \
+        if (type != this->type())                                   \
+            AF_THROW_ERR("Requested type doesn't match array type", \
```
Comment (umar456) [inline src/api/cpp/array.cpp]: `TYPE_ERROR(0, type);`?

File: test/array.cpp [hunk 2]
```diff
@@ -520,3 +520,16 @@ TYPED_TEST(Array, Scalar)
 
     EXPECT_EQ(true, gold[0]==a.scalar<TypeParam>());
 }
+
+TEST(Array, ScalarTypeMismatch)
+{
+    if (noDoubleTests<float>()) return;
+
+    array a = randu(dim4(1), f32);
```
Comment (umar456) [inline test/array.cpp]: Nit: Maybe faster to do a constant or create an af::array from data.

File: test/array.cpp [hunk 3]
```diff
@@ -520,3 +520,16 @@ TYPED_TEST(Array, Scalar)
...
+
+    std::vector<float> gold(a.elements());
```
Comment (umar456) [inline test/array.cpp]: No need to create a vector for one value.

File: src/api/cpp/array.cpp (hunk 1 above)
Comment (9prady9) [reply src/api/cpp/array.cpp]: There are no arguments for this method right?

**Pull request merged**